Initial multi channel trace file format specification (#841)
OpenSimulationInterface/open-simulation-interface: doc/architecture/trace_file_formats.adoc

For obvious reasons the format was deprecated and fully replaced with the `.osi` binary file format.
This release no longer contains any support for the legacy `.txt` file format.

== Multi channel trace file format

=== Overview

The OSI multi channel trace file format is a binary file format that allows for storing multiple serialized OSI message streams of the same or different types in one trace file, along with additional meta-data, and other related data streams.
Due to the nature of the format, it allows for random access to the data streams, and is suitable for more advanced use-cases.

The OSI multi channel trace file format is based on the MCAP file format, which is a generic multi channel trace file format.
The OSI multi channel trace file format is a specialization of the MCAP file format, with additional constraints and requirements specific to OSI.
Hence, any valid OSI multi channel trace file is also a valid MCAP file, but not the other way around.

The following rules apply to OSI multi channel trace files:

- The file extension to be used is `.mcap`.

> - The file must be a valid MCAP file according to the https://mcap.dev/spec[MCAP format specification] version `0x30`.

Author: The spec goes about it in a very round-about way:

> **Magic**
> An MCAP file must begin and end with the following [magic bytes](https://en.wikipedia.org/wiki/File_format#Magic_number):

> 
> `0x89, M, C, A, P, 0x30, \r, \n`

> 
> The byte following "MCAP" is the major version byte. **0x30 is the ASCII character 0.** **Any changes to this specification document (i.e. adding fields to records, introducing new records) will be binary backward-compatible within the major version.**

(Emphasis by me)

So it is stated quite unclearly, but I would treat this as the major version of the specification, given the statements.

> - The file must be a valid MCAP file according to the https://mcap.dev/spec[MCAP format specification] version `0x30`.

Reviewer: That is what confuse me. "include the major version like this" but not "this spec IS major version 0" ...

- Chunk records must either be uncompressed or compressed using either the `zstd` or `lz4` compression algorithms.
- The file must contain a meta-data record with the OSI trace file meta-data defined in section <<sec-osi-trace-file-multi-global-meta-data>>.
  This meta-data record identifies the file as an OSI multi channel trace file.
- The file must contain a schema record in the summary section for each top-level message type that is used in one or more OSI channels with the OSI message schema as defined in section <<sec-osi-trace-file-multi-schema-record>>.
- The file must contain at least one OSI message stream in a channel as defined in section <<sec-osi-trace-file-multi-channel>>.
- The file may contain additional non-OSI message streams in other channels.

NOTE: All OSI and non-OSI message streams stored in the same file share a common time base.
Storing of unrelated message streams in one trace file is therefore not generally useful and should be avoided.

[#sec-osi-trace-file-multi-global-meta-data]
=== Multi channel trace file global meta-data

> The file must contain a meta-data record with the `name` field being `net.asam.osi.trace` and the following mandatory `metadata` entries:

Reviewer 2: ```suggestion
The file must contain a meta-data record with the `name` field being `net.asam.osi.trace`. This name must be unique, and no other meta-data record with the same name is permitted. The record must also include the following mandatory metadata entries:
```

One can easily add metadata with the same name twice. 

```bash
mcap list metadata /tmp/sv_example.mcap 
name              	offset	length	metadata                                                                                                                                 	
net.asam.osi.trace	46    	181   	{"max_osi_version":"3.7.0","max_protobuf_version":"3.21.12","min_osi_version":"3.7.0","min_protobuf_version":"3.21.12","version":"3.7.0"}	
net.asam.osi.trace	227   	122   	{"description":"Example mcap trace file created with the ASAM OSI utilities library."}   

```

Reviewer: That is the reason why the specification explicitly forbids it :)

Author: ```suggestion
The file must contain exactly one meta-data record with the `name` field being `net.asam.osi.trace` and the following mandatory `metadata` entries:
```

`version`::
The semantic version number of the OSI release that this OSI trace file conforms to.
This is a string in the format `major.minor.patch`, for example `3.0.0`.
Note that this version number is not necessarily the same as the OSI schema version(s) used in the trace file.
Rather, it indicates the version of the OSI trace file format itself.

`min_osi_version`::
The minimum version of the OSI schema used in the trace file OSI channels.
This is a string in the format `major.minor.patch`, for example `3.4.0`.

`max_osi_version`::
The maximum version of the OSI schema used in the trace file OSI channels.
This is a string in the format `major.minor.patch`, for example `3.7.0`.

`min_protobuf_version`::
The version of the protobuf implementation used in the trace file.
This is a string in the format `major.minor.patch`, for example `3.17.3`.

`max_protobuf_version`::
The version of the protobuf implementation used in the trace file.
This is a string in the format `major.minor.patch`, for example `3.26.0`.

> `zero_time`::

Reviewer: @pmai I asked around and we have to make this optional/recommended

Argumentation:
1) Naming convention contains it currently and IS optional
2) OSI optional TimeOfDay time_of_day = 2; and optional int64 unix_timestamp = 8; are optional in environment.proto

This legacy format did not contain plain-text, but rather binary protobuf messages separated by a special separator.
For obvious reasons the format was deprecated and fully replaced with the `.osi` binary file format.
This release no longer contains any support for the legacy `.txt` file format.

== Multi channel trace file format

=== Overview

The OSI multi channel trace file format is a binary file format that allows for storing multiple serialized OSI message streams of the same or different types in one trace file, along with additional meta-data, and other related data streams.
Due to the nature of the format, it allows for random access to the data streams, and is suitable for more advanced use-cases.

The OSI multi channel trace file format is based on the MCAP file format, which is a generic multi channel trace file format.
The OSI multi channel trace file format is a specialization of the MCAP file format, with additional constraints and requirements specific to OSI.
Hence, any valid OSI multi channel trace file is also a valid MCAP file, but not the other way around.

> 

Reviewer 3: ```suggestion

The OSI multi channel trace file format does not specify an MCAP profile.
Therefore, the use of the profile record in the MCAP header section is not prescribed.

```

Author: Maybe more to the point:

```suggestion
The OSI multi channel trace file format does not specify an MCAP profile.
Therefore, it can be used with any or no profile record in the MCAP header section, as long as the given profile does not contradict or disallow any of the requirements of this specification.
```

The file must contain a meta-data record with the `name` field being `net.asam.osi.trace` and the following mandatory `metadata` entries:

`version`::
The semantic version number of the OSI release that this OSI trace file conforms to.
This is a string in the format `major.minor.patch`, for example `3.0.0`.
Note that this version number is not necessarily the same as the OSI schema version(s) used in the trace file.
Rather, it indicates the version of the OSI trace file format itself.

`min_osi_version`::
The minimum version of the OSI schema used in the trace file OSI channels.
This is a string in the format `major.minor.patch`, for example `3.0.0`.

`max_osi_version`::
The maximum version of the OSI schema used in the trace file OSI channels.
This is a string in the format `major.minor.patch`, for example `3.0.0`.

`min_protobuf_version`::
The version of the protobuf implementation used in the trace file.
This is a string in the format `major.minor.patch`, for example `3.17.3`.

`max_protobuf_version`::
The version of the protobuf implementation used in the trace file.
This is a string in the format `major.minor.patch`, for example `3.17.3`.

`zero_time`::

> The point in time corresponding to time 0 in all timestamps in the trace file.

Author: ```suggestion
The point in real or simulated calendar time corresponding to time 0 in all timestamps in the trace file.
```

Due to the nature of the format, it allows for random access to the data streams, and is suitable for more advanced use-cases.

The OSI multi channel trace file format is based on the MCAP file format, which is a generic multi channel trace file format.
The OSI multi channel trace file format is a specialization of the MCAP file format, with additional constraints and requirements specific to OSI.
Hence, any valid OSI multi channel trace file is also a valid MCAP file, but not the other way around.

The OSI multi channel trace file format does not specify an MCAP profile.
Therefore, it can be used with any or an empty profile field in the MCAP header record, as long as the given profile does not contradict or disallow any of the requirements of this specification.

The following rules apply to OSI multi channel trace files:

- The file extension to be used is `.mcap`.
- The file must be a valid MCAP file according to the https://mcap.dev/spec[MCAP format specification] version `0x30`.

> - The file must be an indexed MCAP file, including chunk index records in the summary section, with all message records written into chunk records.

Reviewer 4: An editorial note: I think it should be 'shall' instead of 'must'. This applies to most, if not all, uses of the word here.

> [#top-osi_trace_file_formats]

Reviewer 2: Now that there is an issue for that we can resolve this converstation to reduce the effect in one PR  https://github.com/OpenSimulationInterface/open-simulation-interface/issues/842

= OSI trace file formats

== Single channel trace file formats
There are two formats for storing and reading multiple serialized OSI messages of the same type in one trace file.
These formats are very bare bones and do not contain any additional information, like meta-data, schemas, nor do they allow for random access.
For more advanced use-cases, consider using the multi channel trace file format.

*.osi::

> Binary trace file.

Author: ```suggestion
Single channel binary trace file.
```

*.txth::

> Human-readable plain-text trace file.

Author: ```suggestion
Single channel human-readable plain-text trace file.
```

It is strongly recommended to follow reverse domain name notation for custom meta-data record names to avoid conflicts.

[#sec-osi-trace-file-multi-schema-record]
=== OSI message schema

For each OSI top-level message type that is used in one or more OSI channels, the OSI multi channel trace file must contain a corresponding schema record in the summary section.
Note that if multiple versions of the OSI schema are used in the same trace file, a schema record must be included for each version, with different schema IDs.

The schema record must contain the following fields:

`id`::
A file-wide unique non-zero identifier for the schema record.

`name`::
The fully qualified name to the message within the OSI descriptor set. For example, for the `SensorView` message type, this would be `osi3.SensorView`.

`encoding`::
The value `protobuf`.

`data`::
A binary FileDescriptorSet as produced by `protoc --include_imports --descriptor_set_out`.

The schema record must be stored in the summary section of the trace file, and must be referenced by the OSI channels that use the schema.

[#sec-osi-trace-file-multi-channel]
=== OSI channel

> An OSI channel is a data stream within the OSI multi channel trace file that contains serialized OSI messages of the same type.

Author: ```suggestion
An OSI channel is a data stream within the OSI multi channel trace file that contains serialized OSI top-level messages of the same type.
Note that non-top-level messages must not be stored directly in OSI channels.
```

*.txth::
Human-readable plain-text trace file.
Messages are separated by newlines.

> Each message is a serialized OSI message in JSON format.

Author: Good catch, that was a think-o on my part:

```suggestion
Each message is a serialized OSI message in protocol buffer text format.
```

The file must contain a meta-data record with the `name` field being `net.asam.osi.trace` and the following mandatory `metadata` entries:

`version`::
The semantic version number of the OSI release that this OSI trace file conforms to.
This is a string in the format `major.minor.patch`, for example `3.0.0`.
Note that this version number is not necessarily the same as the OSI schema version(s) used in the trace file.
Rather, it indicates the version of the OSI trace file format itself.

`min_osi_version`::
The minimum version of the OSI schema used in the trace file OSI channels.
This is a string in the format `major.minor.patch`, for example `3.0.0`.

`max_osi_version`::
The maximum version of the OSI schema used in the trace file OSI channels.
This is a string in the format `major.minor.patch`, for example `3.0.0`.

`min_protobuf_version`::
The version of the protobuf implementation used in the trace file.
This is a string in the format `major.minor.patch`, for example `3.17.3`.

`max_protobuf_version`::
The version of the protobuf implementation used in the trace file.
This is a string in the format `major.minor.patch`, for example `3.17.3`.

> `zero_time`::

Reviewer 5: I would put this in the optional section, as for many simulation use-cases I don't see a mandatory relation to real-time.

Author: Given that OSI messages contain (and system components use) fields that reference simulated real time all over the place, there is usually always a relationship to some simulated real time, even in pure simulation use cases. Similarly for anything GPS related you want to have valid timestamps, and to even map GPS time to UTC you need the date.

So I would venture that only very rare cases will get away with not having any time relationship (and would have to be validated to ensure there is no missed effects), and hence this should be recorded here. I.e. for simulations that have no particular focus on time effects, usually a convenient fixed time is chosen to correllate with time zero.

In any case this being part of the mandatory meta-data reflects Timm's PR, where it is mandatory as well.

Reviewer 5: see below.

The point in time corresponding to time 0 in all timestamps in the trace file.
Must be provided as a full ISO8601 formatted date time string, including timezone data, conforming to the https://www.w3.org/TR/xmlschema11-2/#dateTimeStamp[XML Schema dateTimeStamp] lexical space.
Values must match the following regular expression:
`-?([1-9][0-9]{3,}|0[0-9]{3})-(0[1-9]|1[0-2])-(0[1-9]|[12][0-9]|3[01])T(([01][0-9]|2[0-3]):[0-5][0-9]:[0-5][0-9](\.[0-9]+)?|(24:00:00(\.0+)?))(Z|(\+|-)((0[0-9]|1[0-3]):[0-5][0-9]|14:00))`
+

> NOTE: that even in pure simulation use cases there is usually a relationship to real time, as on-board components and environment simulation have necessary relationships to real time (for example, the embedded `HostVehicleData` will carry relevant real time information, for localisation and other purposes).

Reviewer 5: ```suggestion
NOTE: that even in pure simulation use cases there is usually a relationship to real time, as on-board components and environment simulation have necessary relationships to real time (for example, the embedded `HostVehicleData` will carry relevant real time information, for localization and other purposes).
```

`creation_time`::

> The point in time when the trace file was created.

Reviewer 5: Is this the beginning or end of writing the trace file?

Author: I don't know, this is taken from Timm's definitions. Personally I would actually not care very much either way, as without additional recording related meta-data (when, how why, with which equipment was this recorded) I don't see very much use for this field for real measurements, and for simulation use the file system timestamps are usually sufficient.

Reviewer 5: Then I propose to put the end time in there, when the mcap file was finalized.

Author: Finalized time stamp would however mean that the meta-data record can only be written at the end of recording (at the end of the data section, as metadata records are not allowed in the summary section at the end). Not impossible to do, but has this effect - start of recording would mean the meta-data record can be written at the beginning (not that this matters all that much, just to keep in mind in discussions).

Reviewer 2: Other tools e.g. ros also use the start of the recording. Nonetheless, I would not name it recording as OSI is mainly focusing on simulations (and also enabling real measurements) where it is simply "writing" but not "recording".

Reviewer 5: I don't have a strong opinion on this. We just need to specify it. We could also put the start of writing to the file as creation_time.

It is strongly recommended to follow reverse domain name notation for custom meta-data keys to avoid conflicts.
The following mandatory entries are defined by {THIS_STANDARD}:

`net.asam.osi.trace.channel.osi_version`:::
The version of the OSI schema used in creating the data of this OSI channel.
This is a string in the format `major.minor.patch`, for example `3.0.0`.

`net.asam.osi.trace.channel.protobuf_version`:::
The version of the protobuf implementation used in creating the data of this OSI channel.
This is a string in the format `major.minor.patch`, for example `3.17.3`.

The following recommended entries are defined by {THIS_STANDARD}:

`net.asam.osi.trace.channel.description`:::
A human-readable description of the channel.

The channel record must be stored in the summary section of the trace file, and must be referenced by the OSI message records that are part of the channel.

All messages in an OSI channel must be stored in chunk records in the data section of the trace file.

Each message record in a chunk record must contain the following fields:

`channel_id`::
The ID of the channel that the message belongs to.

`publish_time`::

> The point in time when the message was published.

Reviewer 5: In #833 we defined the publish_time as:

```
Must reflect the timestamp of the stored OSI top-level message
```

This makes it easier to use the indexing to access certain OSI time frames.

Author: > In #833 we defined the publish_time as:

Actually this was a copy and paste left over, see actual publish_time definition below (will remove this old duplicate).
> 
> ```
> Must reflect the timestamp of the stored OSI top-level message
> ```
> 
> This makes it easier to use the indexing to access certain OSI time frames.

It sadly does not, as solely `log_time` is used for indexing purposes in MCAP, see contortions below with `log_time` to deal with that.

Reviewer 5: Yes, but if we want to put the same values in publish_time and log_time, I propose to make them both identical with the timestamp of the message.

Author: That's what the current definition does, with special allowance to still use log_time as intended if there is a - perceived - need.

> 
> `publish_time`::
> The point in time when the message was published.

Author: ```suggestion
```
